### PR TITLE
implement `sudo --list`

### PR DESCRIPTION
**Pull request by japaric**
minus printing the sudoers policy when called without a command -- that's tracked in #710 

aside from that: permissions validation, PAM authentication, path resolution and credential caching have been implemented and tested.

### CODE REVIEW

File: src/sudo/pipeline/list.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                target_user: &context.target_user,
+                target_group: &context.target_group,
+            };
+            let judgement = pre.check_list_permission(other_user, &context.hostname, list_request);
+
+            if let Authorization::Forbidden = judgement.authorization() {
+                let command = if original_command.is_none() {
+                    "list".to_string()
+                } else {
+                    format!("list{}", context.command.command.display())
```
Comment (rnijveld) [inline src/sudo/pipeline/list.rs:123]: I don't really understand what the point is of emulating this sudo behavior? It just seems very weird to me.

File: src/sudo/pipeline/list.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    pam::CLIConverser,
+    sudo::{pam::PamAuthenticator, SudoersPolicy},
+    sudoers::{Authorization, ListRequest, Policy, Request},
+    system::{timestamp::RecordScope, Process, User},
+};
+
+use super::{AuthPlugin, Pipeline, PolicyPlugin};
+
+impl Pipeline<SudoersPolicy, PamAuthenticator<CLIConverser>> {
+    pub(in crate::sudo) fn run_list(mut self, cmd_opts: SudoOptions) -> Result<(), Error> {
```
[the remark this answers is not in the document]
Comment (japaric) [reply src/sudo/pipeline/list.rs:15]: `Pipeline::run_list` is different enough from the other two `run` methods that it warrants being in its own module, IMO. that way, any helper function / method that it uses, and the other two `run`s won't, can be kept private to that module (see the updated `list.rs` file after a refactor for readability). conversely, code shared between `run_list` and the other `run` methods can go in the parent module (`pipeline.rs`) so that it can be seen by all its users (e.g. see the new `Pipeline::auth_and_update_record_file` method)

^ that approach also avoids files getting too large (`pipeline/list.rs` is ~200 lines; `pipeline.rs` is ~300 lines). I personally prefer to have several self contained files, e.g. one file per feature with an inline `mod tests`, but I know some people find hard to navigate code bases that have many files. (TBH, navigation shouldn't be an issue if you are familiar with LSP's "go to definition / references / type / implementations / etc." actions)

File: src/sudo/pipeline/list.rs (hunk 1 above)
Comment (japaric) [reply src/sudo/pipeline/list.rs]: it was simpler that coming up with a different format. "simpler" as in the compliance tests didn't need any change. diagnostics are user facing / not machine readable so they can be changed at any time without that being consider a (semver) breaking change.

File: src/sudo/pipeline.rs [hunk 3]
```diff
@@ -60,31 +61,12 @@ impl<Policy: PolicyPlugin, Auth: AuthPlugin> Pipeline<Policy, Auth> {
             } => {
                 self.apply_policy_to_context(&mut context, &policy)?;
 
-                let mut auth_status = determine_auth_status(
+                self.auth_and_update_record_file(
```
Comment (japaric) [inline src/sudo/pipeline.rs:64]: the fact that all the fields of `Authorization::Allowed` are passed into this method to me is an indication that those fields should be their own struct, e.g. `enum Authorization { Allowed(AuthorizationAllowed), /* .. */ }` but that feels out of scope for this PR so I won't refactor that in this PR.

Comment (squell) [reply src/sudo/pipeline.rs:64]: Had a similar feeling a while back.

File: src/sudo/pipeline/list.rs (hunk 1 above)
Comment (squell) [reply src/sudo/pipeline/list.rs]: Note that `list` is a pretty recent addition to sudo (I think it was first part of a sudo release after the kick-off of this project), so I feel we have a large degree in freedom in how far to support it (even decide to not support it if we want), based on the criteria we discussed in Berlin (since it's not in wide common usage at the moment).

Anyway, the way original sudo displays this error message feels like it may simply be a mistake in some error formatting function in original sudo that we should communicate to @millert.

For us, if we go for the "it's simpler for now" solution in this PR, we should make an issue to address this error message so we don't forget to fix it later.

Comment (millert) [reply src/sudo/pipeline/list.rs]: Can you elaborate on the sudo error message you are talking about?

**Review by pvdrz — APPROVED** [review]

File: src/common/error.rs [hunk 4]
```diff
@@ -1,12 +1,22 @@
 use crate::pam::PamError;
-use std::{fmt, path::PathBuf};
+use std::{borrow::Cow, fmt, path::PathBuf};
 
 #[derive(Debug)]
 pub enum Error {
+    Silent,
```
Comment (pvdrz) [inline src/common/error.rs:6]: Just curious, why do we need this?

Comment (japaric) [reply src/common/error.rs:6]: `sudo --list some-command` triggers this error path when the invoking user cannot run `sudo some-command` due to sudoers policy. ogsudo fails silently with exit code 1 in that case and we are doing the same right now but we could improve diagnostics in the future and e.g. report back "user ferris cannot run `sudo some-command`( as other-user) on some-hostname`. That improvement would involve repurposing this error variant.

tl;dr it has been added to allow improving diagnostics in the future (if we decide to do so)
